app/models/git_repository.rb

```diff
@@ -114,6 +114,27 @@ def file_changed?(sha1, sha2, file)
     executor.execute!("cd #{pwd}", "git diff --quiet --name-only #{sha1}..#{sha2} #{file}")
   end
 
+  def directory_contents(git_ref, path: '.')
+    cmd = "git ls-tree -r #{git_ref} --name-only --full-tree #{path}"
+    valid, output = run_single_command(cmd) { |line|
+      if block_given?
+        yield line.chomp
+      else
+        line.chomp
+      end
+    }
+    Rails.logger.error("Could not list the files in directory '#{path}'. Git Ref '#{git_ref}'") unless valid
+    output
+  end
+
+  def file_contents(git_ref, file_path)
+    #first command disables less as default pager
+    cmds = ["git config core.pager ''", "git show -q #{git_ref}:#{file_path}"]
+    valid, output = run_commands(cmds) { |line| line }
+    Rails.logger.error("Could not retrieve file contents for '#{file_path}'. Git Ref '#{git_ref}'") unless valid
+    output.join
+  end
```

Comment:
Hummm, that's a good point. I missed that workflow in the previous method as well: `directory_contents`. 
As soon as I fix that workflow, I can run some benchmarks. I actually started by using the Github API, through the `Build#file_from_repo` method you mentioned. But later realised that the local cached repo should already be there, and ended up switching over to this implementation. :)

Thanks for the heads up. I'll let you know if there's any big differences.